[Proposal] Update mediaCapabilitiesProber API

peaBerberian opened the pull request:
After some brainstorming related to #1470, we noticed that as is, our old `mediaCapabilitiesProber` experimental tool could gain by being more flexible.

Especially, the `getCompatibleDRMConfigurations` method checked multiple EME `MediaKeySystemConfiguration` at once and returned information on actual configurations compatible to them once all of them have been checked.

We wondered whether this method would not be more useful by making it less powerful: we'll now only check for a single `MediaKeySystemConfiguration` per method call (and rename that method to `checkDrmConfiguration`).
Doing this signals to an application that it can just combine multiple calls if it wants the old behavior (e.g. through a `Promise.all`), use `Promise.race` to just obtain the first one etc. (this would still be possible with the previous API, but seemed unnatural to application developers).

I also added a `timeout` argument to it, but realized that it might be very complex to maintain, as theoretically `checkDrmConfiguration` could call multiple browser API sequentially in the future (as other `mediaCapabilitiesProber`'s methods are doing).
In that case, where are we supposed to apply the `timeout`: to each browser API call or globally on the method?

---

I also profited from this work to rewrite the implementation of other `mediaCapabilitiesProber`'s  method without changing their API, as I found that a lot of its data-processing code was unneeded and difficult to follow.

peaBerberian committed:
MCP: addCheckDrmConfiguration API

peaBerberian committed:
MCP: Remove confusing log methods

peaBerberian committed:
MCP: rewrite getStatusForHDCP method

peaBerberian committed:
MCP: simplify checkDrmConfiguration implementation

peaBerberian committed:
MCP: re-implement getDecodingCapabilities

peaBerberian committed:
MCP: Remove getCompatibleDRMConfigurations

peaBerberian committed:
MCP: Rewrite getDisplayCapabilities

peaBerberian committed:
MCP: Remove now unneeded code

peaBerberian committed:
Revert "MCP: Remove confusing log methods"

This reverts commit fdd172ad3db7700cc8856a1ab0bcb13610a40204.

peaBerberian committed:
Update MediaCapabilitiesProber documentation

peaBerberian committed:
MCP: Fix unit tests and move around files

peaBerberian committed:
MCP: Remove too complex to maintain timeout option

peaBerberian committed:
MCP: fix integration tests

peaBerberian committed:
Last unit test fix

Florent-Bouisset commented on doc/api/Tools/MediaCapabilitiesProber.md:
```diff
@@ -144,40 +122,26 @@ const mksConfiguration = {
+  ).then(
+  // On success:
+  (config) => {
+    console.log("This device is compatible with the given configuration!");
+    console.log("Wanted configuration:", configuration);
+    console.log("Compatible configuration:", compatibleConfiguration);
```
this example is not working, `configuration` and `compatibleConfiguration` are not defined

peaBerberian replied:
You're right, it's directly the compatible one instead. I'll update.

peaBerberian replied:
done

Florent-Bouisset commented on src/experimental/tools/mediaCapabilitiesProber/probers/isTypeSupportedWithFeatures.ts:
```diff
@@ -14,15 +14,101 @@
+
+  const features = formatTypeSupportedWithFeaturesConfigForAPI(config);
+
+  const result: ISupportWithFeatures = (
+    globalScope as IGlobalScopeWithMSMediaKeysFeatures
+  ).MSMediaKeys.isTypeSupportedWithFeatures(type, features);
```
I'm not sure how this API works, do you have any documentation on this one?

peaBerberian replied:
It was an old edge feature only present on `MSMediaKeys` and was able to check for more precize policy support.

I also have trouble finding documentation on it today, my searches for now only lead to the RxPlayer and complaints from Netflix users :p

peaBerberian replied:
TypeScript at some point had it in its type definitions: https://raw.githubusercontent.com/Microsoft/TypeScript/v2.9.2/lib/lib.es6.d.ts

Though not anymore

Florent-Bouisset commented on src/experimental/tools/mediaCapabilitiesProber/probers/HDCPPolicy.ts:
```diff
@@ -63,41 +56,32 @@ export default function probeHDCPPolicy(
+    }
+  } catch (err) {
+    const error = err instanceof Error ? err : new Error("Unknown Error");
+    log.error("MCP: `probeHDCPPolicy` didn't succeed to create a MediaKeys", error);
```
From what I read, the error log here is not accurate:
We can get here if the API `getStatusForPolicy` does not exist, not because the MediaKeys failed to be created.

peaBerberian replied:
You're right `requestMediaKeySystemAccess` and `createMediaKeys` don't even have their try/catch.

I'll fix that

peaBerberian replied:
done

Florent-Bouisset commented on src/experimental/tools/mediaCapabilitiesProber/probers/__tests__/decodingInfo.test.ts:
```diff
@@ -52,28 +51,27 @@ describe("MediaCapabilitiesProber probers - decodingInfo", () => {
-        resetDecodingInfos();
-      });
+    let thrownException = false;
+    try {
+      await probeDecodingInfos(configuration);
+      resetDecodingInfos();
+    } catch (error) {
+      const message = error instanceof Error ? error.message : "";
+      thrownException = true;
+      expect(message).toEqual("Not enough arguments for calling mediaCapabilites.");
+      expect(decodingInfoStub).not.toHaveBeenCalled();
+      resetDecodingInfos();
+    }
```
remark about `resetDecodingInfos()`,  it can be done only once outside the try..catch

```js
try {
  await probeDecodingInfos(configuration);
} catch (error) {
  const message = error instanceof Error ? error.message : "";
  thrownException = true;
  expect(message).toEqual("Not enough arguments for calling mediaCapabilites.");
  expect(decodingInfoStub).not.toHaveBeenCalled();
}
resetDecodingInfos();
```

But a similar code is runned in the `afterEach()` hook. 
So we can delete that code from every test since it's already done in the hook and check that it works well

peaBerberian replied:
done

Florent-Bouisset commented on src/experimental/tools/mediaCapabilitiesProber/probers/__tests__/decodingInfo.test.ts:
```diff
@@ -82,22 +80,21 @@ describe("MediaCapabilitiesProber probers - decodingInfo", () => {
       },
     };
 
-    expect.assertions(2);
```
I just learned about `expect.assertions()` is it something that should be avoided?

peaBerberian replied:
I don't like using it because it's something easy to forget about and it's not really a meaningful test beside ensuring that a code path with assertions has been taken, in which case I prefer a more explicit check like a boolean.

Florent-Bouisset commented on src/experimental/tools/mediaCapabilitiesProber/probers/__tests__/decodingInfo.test.ts:
```diff
@@ -106,69 +103,73 @@ describe("MediaCapabilitiesProber probers - decodingInfo", () => {
-    // eslint-disable-next-line @typescript-eslint/no-floating-promises
-    expect(probeDecodingInfos({})).rejects.toThrowError(
-      "MediaCapabilitiesProber >>> API_CALL: MediaCapabilities API not available",
+    return probeDecodingInfos({}).then(
```
why is there a `return` ?
Also there is jest modifier `.rejects` that help with this kind of test:
```js
await expect(probeDecodingInfos({})).rejects.toThrowError("navigator.mediaCapabilites.decodingInfo is not available");
```

peaBerberian replied:
When tests return a promise in most testing frameworks, the test will reject when the promise rejects and resolve when the promise resolve.

I know about `rejects`, and that was what was there before, but for some reason I found returning the promise more readable.

Florent-Bouisset commented on src/experimental/tools/mediaCapabilitiesProber/probers/__tests__/decodingInfo.test.ts:
```diff
@@ -182,13 +183,20 @@ describe("MediaCapabilitiesProber probers - decodingInfo", () => {
-    // eslint-disable-next-line @typescript-eslint/no-floating-promises
-    expect(probeDecodingInfos({})).rejects.toThrowError(
-      "MediaCapabilitiesProber >>> API_CALL: Decoding Info not available",
+    return probeDecodingInfos({}).then(
+      () => {
+        throw new Error("Should not succeed");
+      },
+      (err: Error) => {
+        expect(err).toBeInstanceOf(Error);
+        expect(err.message).toEqual(
+          "navigator.mediaCapabilites.decodingInfo is not available",
+        );
+      },
```
Same here

Florent-Bouisset commented on src/experimental/tools/mediaCapabilitiesProber/probers/__tests__/mediaContentType.test.ts:
```diff
@@ -15,10 +14,7 @@ describe("MediaCapabilitiesProber - probers probeMediaContentType", () => {
-      "MediaCapabilitiesProber >>> API_CALL: " + "MediaSource API not available",
-    );
+    expect(() => probeMediaContentType({})).toThrowError("MediaSource API not available");
   });
```
the eslint error could have been fixed by adding `await` before `expect`

peaBerberian replied:
Yes but it's not a promise anymore, so I prefer relying on the more explicit and adapted `toThrowError`

Florent-Bouisset commented on src/experimental/tools/mediaCapabilitiesProber/probers/__tests__/mediaContentType.test.ts:
```diff
@@ -30,9 +26,8 @@ describe("MediaCapabilitiesProber - probers probeMediaContentType", () => {
-    // eslint-disable-next-line @typescript-eslint/no-floating-promises
-    expect(probeMediaContentType({})).rejects.toThrowError(
-      "MediaCapabilitiesProber >>> API_CALL: " + "isTypeSupported not available",
+    expect(() => probeMediaContentType({})).toThrowError(
+      "MediaSource.isTypeSupported API not available",
     );
```
same here

Florent-Bouisset commented on src/experimental/tools/mediaCapabilitiesProber/probers/__tests__/mediaDisplayInfos.test.ts:
```diff
@@ -14,13 +13,8 @@ describe("MediaCapabilitiesProber probers probeMediaDisplayInfos", () => {
-      globalScope as { matchMedia: typeof globalScope.matchMedia | undefined }
-    ).matchMedia = origMatchMedia;
+    expect(() => probeMediaDisplayInfos({})).toThrowError("matchMedia API not available");
+    globalScope.matchMedia = origMatchMedia;
```
setting the `matchMedia` API should be done in the afterEach / beforeEach hooks

peaBerberian replied:
done

Florent-Bouisset commented on src/experimental/tools/mediaCapabilitiesProber/probers/isTypeSupportedWithFeatures.ts:
```diff
@@ -14,15 +14,101 @@
+    support: ISupportWithFeatures,
+  ): "NotSupported" | "Unknown" | "Supported" {
+    if (support === "") {
+      throw new Error("Bad arguments for calling isTypeSupportedWithFeatures");
```
The API returns empty string if arguments are invalid ? Weird one

peaBerberian replied:
Apparently...

Florent-Bouisset commented on src/experimental/tools/mediaCapabilitiesProber/api.ts:
```diff
@@ -0,0 +1,191 @@
+   * @param {Object} mediaConfig
+   * @returns {Promise.<string>}
+   */
+  async getDecodingCapabilities(mediaConfig: IMediaConfiguration): Promise<string> {
```
don't the type should be narrowed to "Supported" / "Unknown" / "NotSupported" instead of being a string ?

peaBerberian replied:
done

peaBerberian committed:
remove unnecessary eslint-disable

peaBerberian committed:
Improve MediaCapabilitiesProber logic after review

Florent-Bouisset commented on src/experimental/tools/mediaCapabilitiesProber/api.ts:
```diff
@@ -0,0 +1,217 @@
+      } catch (err) {
+        log.debug("DRM: KeySystemAccess was granted but it is not usable");
+        return Promise.reject(
+          new Error("Failed to call generateRequest on this keySystem"),
```
Notice that this log can be slightly misleading because the error can be throw from `createMediaKeys` or `createSession` and not exclusively from `generateRequest`

peaBerberian replied:
Yeah, I updated the message for a more generic error and including that error's specific message in it

Florent-Bouisset commented on src/experimental/tools/mediaCapabilitiesProber/api.ts:
```diff
@@ -0,0 +1,217 @@
+   */
+  async getStatusForHDCP(hdcpVersion: string): Promise<string> {
+    if (hdcpVersion === undefined || hdcpVersion.length === 0) {
+      throw new Error("Bad Arguments: " + "No HDCP Policy specified.");
```
the string concatenation is not needed

peaBerberian replied:
Fixed

Florent-Bouisset commented on src/experimental/tools/mediaCapabilitiesProber/api.ts:
```diff
@@ -0,0 +1,217 @@
+          // continue
+          break;
+      }
+    } catch (err) {
```
can be wrote as `catch {...}` without `err` argument, it is not used.
There is multiple occurences in the file.

peaBerberian replied:
I didn't even know that was possible. Turns out it's called "Optional catch binding" and is from [es2019](https://2ality.com/2017/08/optional-catch-binding.html). 

Removed here and in the other places I saw this in this same file.

Florent-Bouisset commented on src/experimental/tools/mediaCapabilitiesProber/probers/HDCPPolicy.ts:
```diff
@@ -29,24 +28,18 @@ export type IMediaKeyStatus =
-export default function probeHDCPPolicy(
-  config: IMediaConfiguration,
-): Promise<[ProberStatus]> {
+export default async function probeHDCPPolicy(
```
I feel that this function can either throw or return "NotSupported"
I'm wondering if we should catch any errors and return "NotSupported"

peaBerberian replied:
Complex to say. 

It fails if EME API failed to be called.

`NotSupported` would be if there's a high priority of the HDCP version being unsupported but here we're not so sure.

Maybe `Unknown`? I'm under the impression that for now `Unknown` is only for when the `getStatusForPolicy` API is not available, which is slightly different than e.g. `requestMediaKeySystemAccess` failing yet both would just indicate that it is unknown.....

Then, when applications handle promises, they generally also handle rejections (even linters often force that). So I'm not sure.

Florent-Bouisset replied:
I removed the Unkown and it only returns "Supported" / "NotSupported" or reject

peaBerberian committed:
mcp: minor updates after review

Florent-Bouisset committed:
prober: throw if the HDCP API is not available

ProbeHDCPPolicy can either, succeed with "Supported" | "NotSupported" | "Unknown" or reject.
Success with "unknown" or reject both occurs when the navigator API failed to execute and the support
of HDCP cannot be determined. To standardize the behavior of this function, the "Unknown"
return is removed and replaced by a promise rejection.

Florent-Bouisset committed:
update doc accordingly to previous change
